nnsvs/mdn.py

```diff
@@ -0,0 +1,169 @@
+        self.out_dim = out_dim
+        self.num_gaussians = num_gaussians
+
+        self.log_pi = nn.Linear(in_dim, num_gaussians)
+        self.log_sigma = nn.Linear(in_dim, out_dim * num_gaussians)
+        self.mu = nn.Linear(in_dim, out_dim * num_gaussians)
+
+    def forward(self, minibatch):
+        log_pi = F.log_softmax(self.log_pi(minibatch), dim=2)
+        log_sigma = self.log_sigma(minibatch)
+        log_sigma = log_sigma.view(len(minibatch), -1, self.num_gaussians, self.out_dim)        
+        mu = self.mu(minibatch)
+        mu = mu.view(len(minibatch), -1, self.num_gaussians, self.out_dim)
+        return log_pi, log_sigma, mu
+
+def mdn_loss(log_pi, log_sigma, mu, target, log_clamp_min=-7.0, reduce=True):
+    """Calculates the error, given the MoG parameters and the target.
+    The loss is the negative log likelihood of the data given the MoG
+    parameters.
+    Arguments:
+        log_pi (B, T, G): The log of multinomial distribution of the Gaussians. B is the batch size,
+            T is data length of this batch, and G is num_gaussians of class MDNLayer.
+        log_sigma (B, T , G ,D_out): The log standard deviation of the Gaussians. D_out is out_dim of class 
+            MDNLayer.
+        mu (B , T, G, D_out): The means of the Gaussians. 
+        target (B, T, D_out): The target variables.
+        log_clamp_min: minimum value of logged values(log_pi, log_sigma, and torch.distributions.Normal.log_prob)
+        reduce: If True, the losses are averaged for each batch.
+    Returns:
+        loss (B) or (B, T): Negative Log Likelihood of Mixture Density Networks.
+    """
+
+    # Clip log_sigma and log_pi with log_clamp_min for numerical stability
+    log_sigma = torch.clamp(log_sigma, min=log_clamp_min)
+    log_pi = torch.clamp(log_pi, min=log_clamp_min)
+    # Expand the dim of target as (B, T, D_out) -> (B, T, 1, D_out) -> (B, T,G, D_out)
+    target = target.unsqueeze(2).expand_as(log_sigma)
+
+    # Create gaussians with mean=mu and variance=torch.exp(log_sigma)^2
+    dist = torch.distributions.Normal(loc=mu, scale=torch.exp(log_sigma))
+
+    # Use torch.log_sum_exp instead of the combination of torch.sum and torch.log
+    # (Reference: https://github.com/r9y9/nnsvs/pull/20#discussion_r495514563)
+
+    # Here we assume that the covariance matrix of multivariate Gaussian distribution is diagonal
+    # to handle the mean and the variance in each dimension separately.
+    # (Reference: https://markusthill.github.io/gaussian-distribution-with-a-diagonal-covariance-matrix/)
+    # log pi(x)N(y|mu(x),sigma(x)) = log pi(x) + log N(y|mu(x),sigma(x))
+    # log N(y_1,y_2,...,y_{D_out}|mu(x),sigma(x)) = log N(y_1|mu(x),sigma(x))...N(y_{D_out}|mu(x),sigma(x))
+    #                                             = \sum_{i=1}^{D_out} log N(y_i|mu(x),sigma(x))
+    # (B, T, G, D_out) -> (B, T, G)
+    log_prob = dist.log_prob(target)
+    log_prob = torch.clamp(log_prob, min=log_clamp_min)
```

Comment:
In my opinion, it would be better to add a clamp before log_prob.

as is: log_prob may produce NaN -> clamp
to be: clamp target not to have NaN in log_prob

What do you think?

Reply:
I'm afraid it may be difficult to clamp target values to avoid nan before log_prob, because the clamping values to be desired may be changed according to the Gaussian distribution.

Reply:
I was thinking about clamping the target values within -5 sigma < x - u < 5 sigma. This a minor issue so it's okay.